### DISCUSSION TITLE
Closures around lexical scope

### DISCUSSION THREAD

**Discussion by ianstranathan** [godotengine/godot-proposals#10280] [Scripting]
Hi all, hopefully I'm posting this in the right place.

Firstly, Godot is awesome and I'm happy to be using it!

Discussion point:
I prefer a more functional style of coding (I know, it's a fools errand as godot is very OOP but...) and sometimes I like
hiding simple things in a function instead of having too many member variables to keep track of (reduces complexity overhead imo)
This seems to be only possible for heap allocated data structures unfortunately (Array, Dict, even Vector2)

A contrived example:
```gdscript
func press_closure(initially_pressed: bool):
	var _is_pressed = initially_pressed
	return func(was_pressed = null) -> bool:
		if was_pressed:
			_is_pressed = was_pressed
			icon_grab(was_pressed)
			#print(_is_pressed)
		# print(_is_pressed)
		return _is_pressed
```

The closure/ lexical scope isn't saved; the returned _is_pressed is always false.
Godot has higher order functions and scope around heap allocated data structures, would it be hard to add proper lexical scope?

Any help or comments is appreciated.
Thanks!

**Comment by theraot** [comment]
See:

- https://github.com/godotengine/godot-proposals/issues/8558

**Reply by ianstranathan** [reply]
Thanks. I tried searching for an already opened, relevant discussion but did not find this.
I've already played with hacking around it using reference types as discussed in link.
Seems like an anti-pattern at this point.
